src/components/common/Modal.tsx

```diff
@@ -0,0 +1,78 @@
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+
+import clsx from 'clsx';
+import { Grow, LinearProgress, Modal as MaterialModal, Paper } from '@material-ui/core';
+
+import usePrevious from '@app/hooks/usePrevious';
+import { ModalOptions, useModal } from '@app/contexts/Modal';
+
+import useStyles from './_jss/Modal.jss';
+
+type ModalProps = Pick<ModalOptions, 'disablePaper' | 'disableIndirectDismissal'>;
+
+interface ModalViewProps {
+  isLoading: boolean;
+  modalProps: ModalProps;
+}
+
+const ModalView: React.FC<ModalViewProps> = ({ children, isLoading, modalProps }) => {
+  const { dismiss, isOpen } = useModal();
+  const [transitionIn, setTransitionIn] = useState(false);
+  const [isVisible, setIsVisible] = useState(isOpen);
+  const lastIsOpen = usePrevious(isOpen);
+  const lastChildren = usePrevious(children);
+  const classes = useStyles();
+
+  const handleClose = useCallback(() => dismiss(true), [dismiss]);
+
+  const isOpening = isOpen && !lastIsOpen;
+  const isClosing = !isOpen && lastIsOpen;
+
+  useEffect(() => {
+    if (isOpening) {
+      setIsVisible(true);
+      setTransitionIn(true);
+    } else if (isClosing) {
+      setTransitionIn(false);
+      setTimeout(() => setIsVisible(false), 400);
+    }
+  }, [isOpening, isClosing]);
```

Comment:
You could just replace this with `isOpen`. It's a primitive value so `isOpen && lastIsOpen` and `!isOpen && !lastIsOpen` will never be true.

Reply:
couldn't, for example, `isOpen && lastIsOpen` be true if the modal is open and the `isLoading` prop changed?

Reply:
Oh yeah I guess it will. My point was more you don't really need `lastOpen`, you could just straight up use `isOpen` in the effect

Reply:
addressing other comments later today but thanks for pointing that out, this helped me think through and clean up this component a lot [`a2149a1`](https://github.com/resident51/resident51/pull/4/commits/a2149a1fe2da5dc7372b9e8bb1fe2db9374aabc8)